### PR TITLE
First steps towards fixing the symlink CVEs

### PR DESCRIPTION
**Pull request by pmatilai**
Details in commits, but basically fixes CVE-2021-35939 and lays down some necessary infrastructure for next steps in securing down our file operations.

### CODE REVIEW

**Comment by pmatilai** [comment]
It should be noted (probably in the commit message too) that as these symlink CVE's overlap and interact in various ways, this does not fully fix CVE-2021-35939 as the directory tracking does not cover all our installation steps yet. Plugging all the holes requires converting all of FSM to the *at() family of calls plus fd-based ops where possible, so this really is just the first step of many to come.

File: INSTALL [hunk 1]
```diff
@@ -103,6 +103,8 @@ option to configure).  For GCC, OpenMP 4.5 is fully supported since GCC 6.1,
 which is available from
     http://www.gnu.org/
 
+Rpm requires a POSIX.1-2008 level operating system.
```
Comment (DemiMarie) [inline INSTALL:106]:
```suggestion
RPM requires a POSIX.1-2008 level operating system.
```

File: lib/fsm.c [hunk 2]
```diff
@@ -406,17 +271,118 @@ static int fsmRmdir(const char *path)
...
 
+static int fsmOpenat(int dirfd, const char *path, int flags)
+{
+    struct stat lsb, sb;
+    int sflags = flags | O_NOFOLLOW;
+    int fd = openat(dirfd, path, sflags);
+
+    /*
+     * Only ever follow symlinks by root or target owner. Since we can't
+     * open the symlink itself, the order matters: we stat the link *after*
```
Comment (DemiMarie) [inline lib/fsm.c:293]: Would Linux `O_PATH` help?

**Review by DemiMarie — COMMENTED** [review]
On Linux, using `openat2()` will be much simpler and more efficient on kernels that support it.  RPM is not Linux-specific, but `openat2()` might be useful where available.

File: lib/fsm.c (hunk 2 above)
Comment (pmatilai) [reply lib/fsm.c:293]: Rpm is not Linux-only.

**Comment by pmatilai** [comment]
I'm quite aware of Linux having all manner of fancy extensions available.
Rpm is portable software and we need to fix this stuff using what's available in POSIX, utilizing non-portable extensions would only make things far more complicated rather than help. There's enough complexities to deal with as it is, thank you very much.

**Comment by pmatilai** [comment]
This is now using fd or dirfd+basename for file ops within the fsm, as much as possible. Plugins pose special problems as external libraries generally dont support dirfd+basename style operation, but may still need to operate on symlinks so we're stuck with "insecure" absolute paths there, for now at least.

I'm seeing a couple of install glitches on fresh chroot install still, but it's getting close now.
Of course a change this big and drastic *will* have bugs in it initially, I have no illusions about that.

**Comment by pmatilai** [comment]
Okay, test-suite + all my local tests (install to empty chroot) pass now :partying_face:

**Comment by DemiMarie** [comment]
> This is now using fd or dirfd+basename for file ops within the fsm, as much as possible. Plugins pose special problems as external libraries generally dont support dirfd+basename style operation, but may still need to operate on symlinks so we're stuck with "insecure" absolute paths there, for now at least.

Cute (but non-portable) trick: use paths of the form `/dev/fd/$FDNUM/something`.  Works at least on Linux.

**Comment by pmatilai** [comment]
Yeah once we have the basics working and optimized to a reasonable degree we can start looking at utilizing various OS-specific extensions. The gotcha with those is to find ways to provide extra functionality in the specific OS'es *without* introducing multiple codepaths (which will inevitably bitrot) to accomplish the same thing.

**Comment by pmatilai** [comment]
Anyway...

There will inevitably be bugs in this all, and since the test-suite covers only so much the best way to find the rest is real-world testing. And sitting in a branch does little to achieve that, so I'm merging this as is now.

Danger Will Robinson, if you're in the habbit of running rpm daily snapshots then you'll want to stay alert for a while.

**Comment by pmatilai** [comment]
Oh, and to make it absolutely clear: we're nowhere *near* done with this, I just want to get this bulk of change over with so we can concentrate with the finer nuances.

**Comment by sandy-lcq** [comment]
@pmatilai  Thanks for fixing these CVEs.  And I want to double check with you that 
does these 32 commits in this pull request fully fix CVE-2021-35937, CVE-2021-35938, CVE-2021-35939? 
Any plan to porting it to 4.17.x branch?

**Comment by pmatilai** [comment]
There will be no backports.

**Comment by Mingli-001** [comment]
Hello, 
Could you help to confirm if I can use the commit https://github.com/rpm-software-management/rpm/commit/25a435e90844ea98fe5eb7bef22c1aecf3a9c033 to fix CVE-2021-35938? Thanks!

**Pull request merged**